File: 2019/5/code.py
import sys

class Interactive:
  def input(self):
    print('input: ')
    return input()
  def output(self, d1):
    print('output: %d' % d1)

class Computer:
  def __init__(self, memory, io):
    self.io = io
    self.running = True
    self.memory = memory
    self.ip = 0
    self.modes = 0

  def log(self, s):
    print(s)

  def quit(self):
    self.log("exiting")
    self.running = False

  def add(self, d1, d2, a3):
    d3 = d1 + d2
    self.memory[a3] = d3
    self.log("[%d] = %d + %d = %d" % (a3, d1, d2, d3))

  def multiply(self, d1, d2, a3):
    d3 = d1 * d2
    self.memory[a3] = d3
    self.log("[%d] = %d * %d = %d" % (a3, d1, d2, d3))

  def input(self, a1):
    d1 = self.io.input()
    self.memory[a1] = d1
    self.log("[%d] = %d" % (a1, d1))

  def output(self, d1):
    self.io.output(d1)

  def jumpif(self, d1, d2):
    if d1 != 0:
      self.ip = d2

  def jumpifnot(self, d1, d2):
    if d1 == 0:
      self.ip = d2

  def lessthan(self, d1, d2, a3):
    if d1 < d2:
      self.memory[a3] = 1
    else:
      self.memory[a3] = 0

  def equals(self, d1, d2, a3):
    if d1 == d2:
      self.memory[a3] = 1
    else:
      self.memory[a3] = 0

  def fetch(self):
    d1 = self.memory[self.ip]
    self.ip = self.ip + 1
    return d1

  def fetchop(self):
    op = self.fetch()
    opcode = op % 100
    op = op / 100
    self.modes = op
    return opcode

  def fetchdata(self):
    mode = self.modes % 10
    self.modes = self.modes / 10
    d1 = self.fetch()
    if mode == 0:
      return self.memory[d1]
    return d1

  def fetchaddr(self):
    mode = self.modes % 10
    self.modes = self.modes / 10
    d1 = self.fetch()
    return d1

  def process(self):
    opcode = self.fetchop()
    if opcode == 99:
      self.quit()
    elif opcode == 1:
      a = self.fetchdata()
      b = self.fetchdata()
      c = self.fetchaddr()
      self.add(a, b, c)
    elif opcode == 2:
      a = self.fetchdata()
      b = self.fetchdata()
      c = self.fetchaddr()
      self.multiply(a, b, c)
    elif opcode == 3:
      a = self.fetchaddr()
      self.input(a)
    elif opcode == 4:
      a = self.fetchdata()
      self.output(a)
    elif opcode == 5:
      a = self.fetchdata()
      b = self.fetchdata()
      self.jumpif(a, b)
    elif opcode == 6:
      a = self.fetchdata()
      b = self.fetchdata()
      self.jumpifnot(a, b)
    elif opcode == 7:
      a = self.fetchdata()
      b = self.fetchdata()
      c = self.fetchaddr()
      self.lessthan(a, b, c)
    elif opcode == 8:
      a = self.fetchdata()
      b = self.fetchdata()
      c = self.fetchaddr()
      self.equals(a, b, c)
    else:
      self.log("unknown opcode: %d" % opcode)
      self.quit()

  def run(self):
    self.log("running %s" % self.memory)
    self.running = True
    self.ip = 0
    while self.running:
      self.process()
      # self.log('state: %s' % self.memory)
    self.log("halted.")
    return self.memory[0]

def run(m):
  print('*' * 80)
  comp = Computer(m, Interactive())
  val = comp.run()
  print('*' * 80)
  return val

def patch(m, n, v):
  m = list(m)
  m[1] = n
  m[2] = v
  return m

def test():
  print(run([1,9,10,3,2,3,11,0,99,30,40,50]))
  print(run([1,0,0,0,99]))
  print(run([2,3,0,3,99]))
  print(run([2,4,4,5,99,0]))
  print(run([1,1,1,4,99,5,6,0,99]))
  print(run([3,0,4,0,99]))
  print(run([1002,4,3,4,33]))
  print(run([1101,100,-1,4,0]))
  print('input 8')
  print(run([3,9,8,9,10,9,4,9,99,-1,8]))
  print('input <8')
  print(run([3,9,7,9,10,9,4,9,99,-1,8]))
  print('input 8')
  print(run([3,3,1108,-1,8,3,4,3,99]))
  print('input <8')
  print(run([3,3,1107,-1,8,3,4,3,99]))
  print('cast to bool')
  print(run([3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9]))
  print('cast to bool')
  print(run([3,3,1105,-1,9,1101,0,0,12,4,12,99,1]))

def main():
  f = file('program.txt')
  s = f.read().strip()
  ss = s.split(',')
  xs = [int(x) for x in ss]
  print(run(xs))

main()

# 1: 4330636
# 2: 10428568

File: 2019/13/code.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type Position struct {
	X int
	Y int
}

type Screen map[Position]int

type Game struct {
	Screen Screen
	Size   Position
	In     <-chan int
	Out    chan<- int
}

func NewGame(in <-chan int) *Game {
	return &Game{
		Screen: make(map[Position]int),
		Size:   Position{X: 0, Y: 0},
		In:     in,
	}
}

func (game *Game) Step() bool {
	x, ok := <-game.In
	if !ok {
		return false
	}
	y, ok := <-game.In
	if !ok {
		return false
	}
	tile, ok := <-game.In
	if !ok {
		return false
	}

	if x > game.Size.X {
		game.Size.X = x
	}
	if y > game.Size.Y {
		game.Size.Y = y
	}

	game.Screen[Position{X: x, Y: y}] = tile
	return true
}

func (game *Game) Run() {
	for game.Step() {
	}
}

func (game *Game) Draw() {
	for row := 0; row <= game.Size.Y; row++ {
		for col := 0; col <= game.Size.X; col++ {
			tile := game.Screen[Position{X: col, Y: row}]
			switch tile {
			case 0:
				fmt.Print(" ")
			case 1:
				fmt.Print("#")
			case 2:
				fmt.Print("+")
			case 3:
				fmt.Print("_")
			case 4:
				fmt.Print("o")
			}
		}
		fmt.Println()
	}
}

func (game *Game) CountBlockTiles() int {
	count := 0
	for _, tile := range game.Screen {
		if tile == 2 {
			count++
		}
	}
	return count
}

type Computer struct {
	In        <-chan int
	Out       chan<- int
	WantInput chan<- bool
	Running   bool
	Memory    map[int]int
	Ip        int
	Modes     int
	Base      int
}

func NewComputer(memory []int, in <-chan int, out chan<- int, wantInput chan<- bool) *Computer {
	mem := make(map[int]int)
	for i, x := range memory {
		mem[i] = x
	}
	return &Computer{
		Memory:    mem,
		In:        in,
		Out:       out,
		WantInput: wantInput,
		Running:   true,
		Ip:        0,
		Modes:     0,
		Base:      0,
	}
}

func (comp *Computer) quit() {
	comp.Running = false
}

func (comp *Computer) add(d1, d2, a3 int) {
	d3 := d1 + d2
	comp.Memory[a3] = d3
}

func (comp *Computer) multiply(d1, d2, a3 int) {
	d3 := d1 * d2
	comp.Memory[a3] = d3
}

func (comp *Computer) input(a1 int) {
	d1 := <-comp.In
	comp.Memory[a1] = d1
}

func (comp *Computer) output(d1 int) {
	comp.Out <- d1
}

func (comp *Computer) jumpIf(d1, d2 int) {
	if d1 != 0 {
		comp.Ip = d2
	}
}

func (comp *Computer) jumpIfNot(d1, d2 int) {
	if d1 == 0 {
		comp.Ip = d2
	}
}

func (comp *Computer) lessThan(d1, d2, a3 int) {
	if d1 < d2 {
		comp.Memory[a3] = 1
	} else {
		comp.Memory[a3] = 0
	}
}

func (comp *Computer) equals(d1, d2, a3 int) {
	if d1 == d2 {
		comp.Memory[a3] = 1
	} else {
		comp.Memory[a3] = 0
	}
}

func (comp *Computer) fetch() int {
	d1 := comp.Memory[comp.Ip]
	comp.Ip++
	return d1
}

func (comp *Computer) fetchOp() int {
	op := comp.fetch()
	opcode := op % 100
	op = op / 100
	comp.Modes = op
	return opcode
}

func (comp *Computer) fetchData() int {
	mode := comp.Modes % 10
	comp.Modes = comp.Modes / 10
	d1 := comp.fetch()
	if mode == 0 {
		return comp.Memory[d1]
	}
	if mode == 1 {
		return d1
	}
	if mode == 2 {
		return comp.Memory[d1+comp.Base]
	}
	panic("unknown data mode")
}

func (comp *Computer) fetchAddr() int {
	mode := comp.Modes % 10
	comp.Modes = comp.Modes / 10
	d1 := comp.fetch()
	if mode == 0 {
		return d1
	}
	if mode == 2 {
		return d1 + comp.Base
	}
	panic("unknown addr mode")
}

func (comp *Computer) process() {
	switch opcode := comp.fetchOp(); opcode {
	case 99:
		comp.quit()
	case 1:
		a := comp.fetchData()
		b := comp.fetchData()
		c := comp.fetchAddr()
		comp.add(a, b, c)
	case 2:
		a := comp.fetchData()
		b := comp.fetchData()
		c := comp.fetchAddr()
		comp.multiply(a, b, c)
	case 3:
		a := comp.fetchAddr()
		comp.input(a)
	case 4:
		a := comp.fetchData()
		comp.output(a)
	case 5:
		a := comp.fetchData()
		b := comp.fetchData()
		comp.jumpIf(a, b)
	case 6:
		a := comp.fetchData()
		b := comp.fetchData()
		comp.jumpIfNot(a, b)
	case 7:
		a := comp.fetchData()
		b := comp.fetchData()
		c := comp.fetchAddr()
		comp.lessThan(a, b, c)
	case 8:
		a := comp.fetchData()
		b := comp.fetchData()
		c := comp.fetchAddr()
		comp.equals(a, b, c)
	case 9:
		a := comp.fetchData()
		comp.Base = comp.Base + a
	default:
		fmt.Printf("unknown opcode: %d\n", opcode)
		comp.quit()
	}
}

func (comp *Computer) Run() {
	comp.Running = true
	comp.Ip = 0
	for comp.Running {
		comp.process()
	}
	close(comp.Out)
	close(comp.WantInput)
}

func ReadProgram() []int {
	b, err := ioutil.ReadFile("program.txt")
	if err != nil {
		panic(err)
	}
	s := strings.TrimSpace(string(b))
	parts := strings.Split(s, ",")
	p := make([]int, len(parts))
	for i, part := range parts {
		p[i], err = strconv.Atoi(part)
		if err != nil {
			panic(err)
		}
	}
	return p
}

func main() {
	p := ReadProgram()
	in := make(chan int)
	out := make(chan int)
	wantInput := make(chan bool)
	done := make(chan bool)

	c := NewComputer(p, in, out, wantInput)
	// c.Memory[0] = 2
	game := NewGame(out)
	go func() {
		c.Run()
		done <- true
	}()
	go func() {
		game.Run()
		done <- true
	}()
	go func() {
		_, ok := <-wantInput
		game.Draw()
		if ok {
			in <- 1
		}
		done <- true
	}()

	<-done
	<-done
	<-done
	close(in)

	fmt.Printf("block tiles: %d\n", game.CountBlockTiles())
}

// 361
